### ISSUE TITLE
Memory leak with dynamic styles

### ISSUE THREAD

**Issue by qhxin** [cssinjs/jss#1360]
__Expected behavior:__
My APP must render very frequently by our realtime data.

__Describe the bug:__
My APP has memory leak on Firefox, after using material-ui. 
This is the same issue in material-ui https://github.com/mui-org/material-ui/issues/21528

__Codesandbox link:__ 
 [https://codesandbox.io/s/dazzling-visvesvaraya-eczqe?file=/src/Tab.js](https://codesandbox.io/s/dazzling-visvesvaraya-eczqe?file=/src/Tab.js) 

__Versions (please complete the following information):__
 - jss: "^10.0.3"
 - Browser [e.g. chrome, safari]:  firefox
 - OS [e.g. Windows, macOS]: macOS
Feel free to add any additional versions which you may think are relevant to the bug.

**Comment by qhxin** [comment]
the leak selector is 
```

          "&:not(:first-child)": {
            borderLeft: "none"
          }
```

**Comment by kof** [comment]
will it happen if you just put this: :
```

          "&": {
            borderLeft: "none"
          }
```

**Comment by kof** [comment]
Can you reproduce it without using react and material-ui? If the assumption is that it's exactly this rule, it means you only need jss core and jss-plugin-nested for reproduction

**Comment by qhxin** [comment]
> will it happen if you just put this?:
> 
> ```
>           "&": {
>               borderLeft: "none"
>           }
> ```

yeah,
```
 "&": {
}
```
this selector is the same issue.

**Comment by qhxin** [comment]
> Can you reproduce it without using react and material-ui? If the assumption is that it's exactly this rule, it means you only need jss core and jss-plugin-nested for reproduction.
> 
> If we don't have a bare minimum for reproduction we are unlikely to fix it.

i reproduced it in this sandbox: https://codesandbox.io/s/strange-hill-z3gwr?file=/index.html
the sheet leak when the dynamic style function update is processing.

**Comment by Jokcy** [comment]
@kof It seems `jss-plugin-nested` add a new Rule to RuleList everytime when update, without delete old rule. It make `RuleList.index` increase evenytime when update. 

I want to help for this. Now I need to know if there is anyway to remove old useless rule when add, or I have to start with it.

**Comment by Jokcy** [comment]
@kof I try to read the code, I can't quite understand about this code:
```js
let key = name
    if (name in this.raw) {
      key = `${name}-d${this.counter++}`
    }

```
when addRule to RuleList, even if the same rule name exist, it just create a new name, this make same rule add to rule list every time. Why we need to do like this?

**Comment by kof** [comment]
@Jokcy right above that line there is the comment:
```
    // When user uses .createStyleSheet(), duplicate names are not possible, but
    // `sheet.addRule()` opens the door for any duplicate rule name. When this happens
    // we need to make the key unique within this RuleList instance scope.
```

Is this not explaining it?

**Comment by Jokcy** [comment]
Yeah, I already read that, my quesstion actually is about it, why we need to keep rules with same name? I just start to read jss code yesterday, so I did not catch a lot of detail now.

**Comment by kof** [comment]
Mainly because as a user you may call sheet.addRule() from where you are not aware if there is another rule with the same name already there, so the idea was to allow doing so while resolving name conflicts automatically.

I wonder if there is no way in our case here to identify the same rule (without stringification)

**Comment by kof** [comment]
Also feel free to locally disable this line and run the tests to see what other use cases will be failing.

**Comment by Jokcy** [comment]
@kof I'm wondering if we can just add another API to fix this. In this API we won't do:

```js
if (name in this.raw) {
    key = `${name}-d${this.counter++}`
}
```

This API just for plugins and people who know this issue. I tested in the codesandbox demo, it did work, but I got some test cases need to fix.

I need some advice to know if  I went into bad direction.

**Comment by naripok** [comment]
Any plans to merge this PR?
Can I help in any way?

**Comment by kof** [comment]
@naripok you are refering to #1381 it would make sense to read the PR and the comments

**Comment by naripok** [comment]
> @naripok you are refering to #1381 it would make sense to read the PR and the comments

Hey @kof,
I'm sorry, I meant to comment on the PR. Did not have coffee yet.
I read the PR and related issues and got the information that it is not merged because the snapshots needs to be updated.
I'm asking about it because I want to avoid double work, but I can give it a try latter today if it is the case.
Didn't mean to be a push over. I'll take a better look latter and see if I can help.

**Comment by qhxin** [comment]
cssinjs/jss would be eliminated by material-ui...

https://github.com/mui-org/material-ui/issues/22342

**Comment by kof** [comment]
https://github.com/cssinjs/jss/pull/1381#issuecomment-731570240

**Comment by seiyab** [comment]
I found that we should take care of following to solve the memory leak.
- Of course, we need to solve memory leak rendered in [CSSStyleSheet](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet)
- And also, we need to solve memory leak in `RuleList.index`, `RuleList.raw`, `RuleList.map`.
- Additionally, we might need to solve memory leak in `DOMRenderer.cssRules`
- We should write tests for all the above memory leaks.

**Comment by kof** [comment]
released in https://github.com/cssinjs/jss/releases/tag/v10.8.1
